### PR TITLE
FM2-394: Add support for $lastn operation on Observations in R3

### PR DESCRIPTION
**Pull request by Medhavi-16**
## Issue I worked on

see https://issues.openmrs.org/browse/FM2-394

## Checklist: I completed these to help reviewers :)

- [ ] My IDE is configured to follow the [**code style**](https://wiki.openmrs.org/display/docs/Java+Conventions) of this project.

  No? Unsure? -> [configure your IDE](https://wiki.openmrs.org/display/docs/How-To+Setup+And+Use+Your+IDE), format the code and add the changes with `git add . && git commit --amend`

- [ ] I have **added tests** to cover my changes. (If you refactored
  existing code that was well tested you do not have to add tests)

  No? -> write tests and add them to this commit `git add . && git commit --amend`

- [ ] I ran `mvn clean package` right before creating this pull request and
  added all formatting changes to my commit.

  No? -> execute above command

- [ ] All new and existing **tests passed**.

  No? -> figure out why and add the fix to your commit. It is your responsibility to make sure your code works.

- [ ] My pull request is **based on the latest changes** of the master branch.

  No? Unsure? -> execute command `git pull --rebase upstream master`

### CODE REVIEW

**Comment by Medhavi-16** [comment]
@ibacher @varung-31 PTAL.

File: api/src/test/java/org/openmrs/module/fhir2/providers/r3/ObservationFhirResourceProviderTest.java [hunk 1]
```diff
@@ -315,4 +323,197 @@ public void deleteObservation_shouldThrowResourceNotFoundExceptionWhenIdRefersTo
...
+		
+		assertThat(results, notNullValue());
+		assertThat(resultList, hasSize(equalTo(1)));
+		assertThat(resultList.get(0), notNullValue());
+		assertThat(resultList.get(0).fhirType(), equalTo(FhirConstants.OBSERVATION));
+		assertThat(resultList.get(0).getIdElement().getIdPart(), equalTo(OBSERVATION_UUID));
+	}
+	
+	@Test
+	public void getLastnObservations_shouldReturnRecentNObservationsWhenNoCodeIsSpecified() {
```
Comment (HerbertYiga) [inline api/src/test/java/org/openmrs/module/fhir2/providers/r3/ObservationFhirResourceProviderTest.java:471]: thanks @Medhavi-16  hi @ibacher  how do things look to you for this entire pr?

File: api/src/test/java/org/openmrs/module/fhir2/providers/r3/ObservationFhirResourceProviderTest.java [hunk 2]
```diff
@@ -315,4 +323,197 @@ public void deleteObservation_shouldThrowResourceNotFoundExceptionWhenIdRefersTo
...
+		referenceParam.addValue(new ReferenceOrListParam().add(patient));
+		
+		TokenAndListParam categories = new TokenAndListParam().addAnd(new TokenParam().setValue("laboratory"));
+		
+		TokenAndListParam code = new TokenAndListParam().addAnd(
+		    new TokenParam().setSystem(FhirTestConstants.LOINC_SYSTEM_URL).setValue(LOINC_SYSTOLIC_BP),
+		    new TokenParam().setSystem(FhirTestConstants.CIEL_SYSTEM_URN).setValue(CIEL_DIASTOLIC_BP));
+		
+		when(observationService.getLastnObservations(max, referenceParam, categories, code)).thenReturn(
+		    new org.openmrs.module.fhir2.providers.r4.MockIBundleProvider<>(Collections.singletonList(observation), 10, 1));
```
Comment (varung-31) [inline api/src/test/java/org/openmrs/module/fhir2/providers/r3/ObservationFhirResourceProviderTest.java:344]: We should use the MockIBundleProvider for R3. Please change this throughout the PR.

File: omod/src/test/java/org/openmrs/module/fhir2/providers/r3/ObservationFhirResourceProviderWebTest.java [hunk 3]
```diff
@@ -794,6 +798,110 @@ public void shouldHandleMultipleReverseIncludes() throws Exception {
...
+
+		TokenOrListParam orListParam = codeCaptor.getValue().getValuesAsQueryTokens().get(0);
+		assertThat(orListParam.getValuesAsQueryTokens(), notNullValue());
+		assertThat(orListParam.getValuesAsQueryTokens().size(), equalTo(1));
+		assertThat(orListParam.getValuesAsQueryTokens().get(0).getSystem(), nullValue());
+		assertThat(orListParam.getValuesAsQueryTokens().get(0).getValue(), equalTo("5085"));
+	}
+	
+	private void verifyLastnOperation(String uri) throws Exception {
+		org.hl7.fhir.r4.model.Observation observation = new org.hl7.fhir.r4.model.Observation();
```
Comment (varung-31) [inline omod/src/test/java/org/openmrs/module/fhir2/providers/r3/ObservationFhirResourceProviderWebTest.java:887]: We should use the R3 Observation resource.

**Review by varung-31 — CHANGES_REQUESTED** [review]
@Medhavi-16, This PR mostly looks good apart from a couple of changes.

File: api/src/test/java/org/openmrs/module/fhir2/providers/r3/ObservationFhirResourceProviderTest.java (hunk 2 above)
Comment (Medhavi-16) [reply api/src/test/java/org/openmrs/module/fhir2/providers/r3/ObservationFhirResourceProviderTest.java]: Yes, I'll do that. I apologise for such mistakes.

**Review by varung-31 — APPROVED** [review]
LGTM!

**Pull request merged**